File: Programmers/가장 먼 노드.cpp
//https://programmers.co.kr/learn/courses/30/lessons/49189?language=cpp

#include <string>
#include <vector>
#include <queue>
using namespace std;

int solution(int n, vector<vector<int>> edge) {
    int answer = 0;
    
    vector<int> v[n+1];   //인접리스트
    int dis[20001] = {0}; //노드 1에서부터 거리
    
    for(int i = 0; i<edge.size(); ++i){  //인접리스트 구현
        v[edge[i][0]].push_back(edge[i][1]);
        v[edge[i][1]].push_back(edge[i][0]);
    }
    
    //bfs
    queue<int> q;
    q.push(1);
    dis[1] = 1;
    int d = 0;
    while(!q.empty()){
        d++;
        int size = q.size();
        while(size--){
            int now = q.front();
            q.pop();
            for(int i = 0; i<v[now].size(); ++i){
                if(dis[v[now][i]]) continue;
                dis[v[now][i]] = d;
                q.push(v[now][i]);
            }
        }
    }
    
    for(int i = 1; i<=n; ++i){
        if(dis[i] == d-1) answer++;
    }
    
    return answer;
}

File: Programmers/가장 먼 노드.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] edge) {
        int answer = 0;
        ArrayList<Integer>[] adj = new ArrayList[n+1];
        int[] dis = new int[n+1];
        
        for(int i = 1; i<n+1; ++i){
            adj[i] = new ArrayList<>();
        }
        
        for(int i = 0; i<edge.length; ++i){
            int a = edge[i][0];
            int b = edge[i][1];
            adj[a].add(b);
            adj[b].add(a);
        }
        
        Queue<Integer> q = new LinkedList<>();
        q.add(1);
        dis[1] = 1;
        int d = 0;
        while(!q.isEmpty()){
            int size = q.size();
            d++;
            while(size!=0){
                size--;
                int now = q.peek();
                q.poll();
                for(int i = 0; i<adj[now].size(); ++i){
                    int next = adj[now].get(i);
                    if(dis[next]!=0) continue;
                    dis[next] = d;
                    q.add(next);
                }
            }
        }
        
        for(int i = 1; i<n+1; ++i){
            if(dis[i] == d-1) answer++;
        }
        
        return answer;
    }
}